### ISSUE TITLE
Better "not 48KHz" error message on MacOS

### ISSUE THREAD

**Issue by chrisrimple** [corrados/jamulus#659]
A new/clean install of Jamulus on MacOS defaults to Device = in:System/out:System.  That's great, except that MacOS defaults to 44.1KHz for the System devices.  It's not hard to fix on MacOS, but most users that are new to Jamulus (1) won't know how and (2) have never used the required application before.  So perhaps, and maybe just on MacOS, a first-launch message is needed to (A) alert the user to this situation and (B) tell them how to fix it (or provide a link to a Wiki page with that information).

In short, the user must:

- Open Applications > Utilities > Audio MIDI Setup
- From the Window menu, select Show Audio Devices
- Select each audio device listed (on the left) and change the Format value to 48KHz (16-bit or 24-bit); this is particularly true for the devices Built-In Microphone and Built-In Output if the user will be using Jamulus with Device = in:System/out:System
- Quit Audio MIDI Setup

**Comment by corrados** [comment]
That should already be implemented. If the sample rate is not correctly set, a dialog with the following content should open automatically:

            return QString ( tr ( "Current system audio input device sample "
                "rate of %1 Hz is not supported. Please open the Audio-MIDI-Setup in "
                "Applications->Utilities and try to set a sample rate of %2 Hz." ) ).arg (
                static_cast<int> ( inputSampleRate ) ).arg ( SYSTEM_SAMPLE_RATE_HZ );

**Comment by chrisrimple** [comment]
Oh, interesting.  I just confirmed this by using a headset that is only capable of 32KHz.

But here's more fun: with Audio MIDI Setup open, I tried setting my Built-In Output to 44.1KHz, then selecting it in Jamulus, and did not get the error.  I didn't get the error because Jamulus appears to be resetting it to 48KHz, which I can see happening "live" in Audio MIDI Setup.  Is Jamulus now "requesting" 48KHz from the device and setting it automatically if available?

**Comment by corrados** [comment]
Yes, that's actually what happens. Here is the above code plus some more lines:

    if ( inputSampleRate != fSystemSampleRate )
    {
        // try to change the sample rate
        if ( AudioObjectSetPropertyData ( audioInputDevice[iDriverIdx],
                                          &stPropertyAddress,
                                          0,
                                          NULL,
                                          sizeof ( Float64 ),
                                          &fSystemSampleRate ) != noErr )
        {
            return QString ( tr ( "Current system audio input device sample "
                "rate of %1 Hz is not supported. Please open the Audio-MIDI-Setup in "
                "Applications->Utilities and try to set a sample rate of %2 Hz." ) ).arg (
                static_cast<int> ( inputSampleRate ) ).arg ( SYSTEM_SAMPLE_RATE_HZ );
        }
    }

As you can see, it checks if the sampling rate is ok. If not it tries to correct it. If that does not work, it will show you the error message.

**Comment by chrisrimple** [comment]
Brilliant.  Then let's close my issue.  :)
